### ISSUE TITLE
rustdoc: generic constant expressions not displayed (and where clauses using said gces)

### ISSUE THREAD

**Issue by bend-n** [rust-lang/rust#120994]
This code

```rust
#![feature(generic_const_exprs)]
pub const fn assert_zero(x: usize) -> usize {
    if x != 0 {
        panic!("expected zero");
    } else {
        0
    }
}

pub fn chunked<const C: usize, const N: usize, T>(x: [u8; N]) -> [[T; C]; N / C]
where
    [(); assert_zero(N % C)]:,
{
    unsafe {
        core::mem::ManuallyDrop::new(x)
            .as_ptr()
            .cast::<[[T; C]; N / C]>()
            .read()
    }
}
```

Produces this documentation

![image](https://github.com/rust-lang/rust/assets/70787919/a1001b0b-6a29-4763-9bd9-3d4a9a6042c5)

This is what i want

![image](https://github.com/rust-lang/rust/assets/70787919/4336b99f-2064-4892-a708-b1278fbf00b3)

### Meta
rustdoc 1.78.0-nightly (1a648b397 2024-02-11)

@GuillaumeGomez @fmease

**Comment by fmease** [comment]
Right, I've anonymized most const exprs in #98814[^1] as a hot fix for #97933 to prevent private/internal const exprs getting leaked, so this kind of intentional although the heuristic is super conservative.

However, this doesn't mean that this will never change. I originally (roughtly 2 years ago) set out to improve rustdoc's const & generic const story (cc #99688). This has stalled a bit but only because I'm super involved in different rustc projects at the moment and I definitely want to pick this up again.

For context, I plan on writing a rustdoc RFC at some point (target date: mid 2024) for richer const “support” in rustdoc (cc #98929, #99630, #102456).

Since everything (the `/` and the `assert_zero`) is public, ideally we would properly display them.
I will gladly mentor & review a PR that splits the current heuristic (which only works for local items anyways) into (1) const exprs as bound to free & assoc const items (2) const exprs as used as a const arg and make (2) consider more things public. The BinOp `/` can definitely be considered public (we don't need to worry about impl restrictions yet, #106074) and we can talk about `assert_zero` (for that we probably need to rewrite the way rustdoc represents const exprs (for local items they are HIR const exprs if I remember correctly which doesn't contain the information if a function is public etc., we would need to use `ty::Const` here).

[^1]: That's probably why you've tagged me.

**Comment by fmease** [comment]
The empty bound not getting displayed I would consider a bug since there's a semantic difference between `fn f<'a, T>()` and `fn f<'a, T>() where 'a:, T:` even outside of `generic_const_exprs`: It affects “boundness” (early vs late) and well-formed checking. I'm pretty sure we currently filter them out. I will gladly accept any PRs fixing this.

**Comment by bend-n** [comment]
If you have `T: Sealed`, it still shows the bound. I dont see why
> Since everything (the / and the assert_zero) is public, ideally we would properly display them.
Whats wrong with it displaying the function name without any link?

**Comment by bend-n** [comment]
Your linked issue seems to be about associated constants, not about types.

**Comment by fmease** [comment]
> Your linked issue seems to be about associated constants, not about types.

Right and the fix was for all (local) const exprs since back then I tarred all with the same brush. However your issue and your further comments does make me realize that we probably don't need to apply the heuristic to const args :+1:

**Comment by fmease** [comment]
Just as an addendum, technically speaking, all constant *values*[^1] regardless of location (const argument, free or assoc const item) are part of the public API and need to adhere to SemVer. E.g. changing `pub const X: i32 = 0;` to `pub const X: i32 = 1;` is a breaking change since it can stop the code of downstream users from compiling if used as const argument. In reality however very few people actually know or care about this.

[^1]: That are of a type that impls `ConstParamTy`

**Comment by fmease** [comment]
> Whats wrong with it displaying the function name without any link?

Because it's kind of hard to come up with a good heuristic when to show and when not to show complex const exprs. For example should we display the whole const expr in `pub fn f(_: I32<{ let x = 0; x }>)`?

**Comment by bend-n** [comment]
So this is largely a design problem, yes?

**Comment by fmease** [comment]
Yes, the implementation is straightforward (unless we want to take extra stuff into consideration like the visibility of the functions used inside the const exprs but that's also solvable).
